base/tuple.jl

```diff
@@ -20,7 +20,10 @@ endof(t::Tuple) = length(t)
 size(t::Tuple, d) = d==1 ? length(t) : throw(ArgumentError("invalid tuple dimension $d"))
 getindex(t::Tuple, i::Int) = getfield(t, i)
 getindex(t::Tuple, i::Real) = getfield(t, convert(Int, i))
-getindex(t::Tuple, r::AbstractArray{<:Any,1}) = tuple([t[ri] for ri in r]...)
+function getindex(t::Tuple, r::AbstractArray{<:Any,1})
+    s = Ref(start(r))
+    ntuple(j -> ((i, s[]) = next(r, s[]); t[i]), length(r))
+end
```

Comment:
Nice find! Inference's sensitivity to such things is somewhat disconcerting. Best!